components/telemetry-controller/controllers/loggingconfiguration_controller.go

```diff
@@ -0,0 +1,327 @@
+
+	var config telemetryv1alpha1.LoggingConfiguration
+	if err := r.Get(ctx, req.NamespacedName, &config); err != nil {
+		log.Error(err, "unable to fetch LoggingConfiguration")
+		// we'll ignore not-found errors, since they can't be fixed by an immediate
+		// requeue (we'll need to wait for a new notification), and we can get them
+		// on deleted requests.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	updatedCm, err := r.syncConfigMap(ctx, &config)
+	if err != nil {
+		log.Error(err, "failed to sync ConfigMap")
+		return ctrl.Result{}, err
+	}
+
+	updatedFiles, err := r.syncFiles(ctx, &config)
+	if err != nil {
+		log.Error(err, "failed to sync mounted files")
+		return ctrl.Result{}, err
+	}
+
+	updatedEnv, err := r.syncEnvironment(ctx, &config)
+	if err != nil {
+		log.Error(err, "failed to sync environment variables")
+		return ctrl.Result{}, err
+	}
+
+	if updatedCm || updatedFiles || updatedEnv {
+		if err := r.Update(ctx, &config); err != nil {
+			log.Error(err, "cannot update LoggingConfiguration")
+			return ctrl.Result{}, err
+		}
+
+		if err := r.deleteFluentBitPods(ctx, log); err != nil {
+			log.Error(err, "cannot delete fluent bit pods")
+			return ctrl.Result{}, err
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *LoggingConfigurationReconciler) syncConfigMap(ctx context.Context, config *telemetryv1alpha1.LoggingConfiguration) (bool, error) {
+	var cm corev1.ConfigMap
+
+	// Get or create ConfigMap
+	if err := r.Get(ctx, r.FluentBitConfigMap, &cm); err != nil {
+		if errors.IsNotFound(err) {
+			cm = corev1.ConfigMap{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      r.FluentBitConfigMap.Name,
+					Namespace: r.FluentBitConfigMap.Namespace,
+				},
+			}
+			if err := r.Create(ctx, &cm); err != nil {
+				return false, err
+			}
+		} else {
+			return false, err
+		}
+	}
+	cmKey := config.Name + ".conf"
+
+	// Add or remove Fluent Bit configuration sections
+	if config.DeletionTimestamp != nil {
+		if cm.Data != nil && controllerutil.ContainsFinalizer(config, configMapFinalizer) {
+			delete(cm.Data, cmKey)
+			controllerutil.RemoveFinalizer(config, configMapFinalizer)
+		}
+	} else {
+		fluentBitConfig := generateFluentBitConfig(config.Spec.Sections)
+		if cm.Data == nil {
+			data := make(map[string]string)
+			data[cmKey] = fluentBitConfig
+			cm.Data = data
+		} else {
+			if oldConfig, hasKey := cm.Data[cmKey]; hasKey && oldConfig == fluentBitConfig {
+				// Nothing changed
+				return false, nil
+			}
+			cm.Data[cmKey] = fluentBitConfig
+		}
+		controllerutil.AddFinalizer(config, configMapFinalizer)
+	}
+
+	// Update ConfigMap
+	if err := r.Update(ctx, &cm); err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (r *LoggingConfigurationReconciler) syncFiles(ctx context.Context, config *telemetryv1alpha1.LoggingConfiguration) (bool, error) {
+	var cm corev1.ConfigMap
+	changed := false
+
+	// Get or create ConfigMap
+	if err := r.Get(ctx, r.FluentBitFilesConfigMap, &cm); err != nil {
```

Comment:
Same code as in the syncConfigMaps function, please consider outsourcing into a function